core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataTest.scala

```diff
@@ -307,6 +310,41 @@ class GroupMetadataTest {
     assertEquals(Some(offset), group.offset(partition))
   }
 
+  @Test
+  def testStableGroupChooseExpiredOffsets(): Unit = {
+    val tp1 = new TopicPartition("unsubscribed_and_expired", 0)
+    val tp2 = new TopicPartition("unsubscribed_and_unexpired", 0)
+    val tp3 = new TopicPartition("subscribed_and_expired", 0)
+    val tp4 = new TopicPartition("subscribed_and_unexpired", 0)
+    val now = Time.SYSTEM.milliseconds
+    val retentionMs = 10000
+    val expiredOffset = OffsetAndMetadata(37, "", now - retentionMs - 1000)
+    val unexpiredOffset = OffsetAndMetadata(37, "", now)
+    group.prepareOffsetCommit(Map(tp1 -> expiredOffset, tp2 -> unexpiredOffset, tp3 -> expiredOffset, tp4 -> unexpiredOffset))
+    group.onOffsetCommitAppend(tp1, CommitRecordMetadataAndOffset(Some(3), expiredOffset))
+    group.onOffsetCommitAppend(tp2, CommitRecordMetadataAndOffset(Some(3), unexpiredOffset))
+    group.onOffsetCommitAppend(tp3, CommitRecordMetadataAndOffset(Some(3), expiredOffset))
+    group.onOffsetCommitAppend(tp4, CommitRecordMetadataAndOffset(Some(3), unexpiredOffset))
+
+    // Set group state to Stable to see whether any expired offsets could be retrieved via `removeExpiredOffsets`
+    group.transitionTo(PreparingRebalance)
+    group.transitionTo(CompletingRebalance)
+    group.transitionTo(Stable)
+
+    val member = new MemberMetadata("memberId", groupId, groupInstanceId, clientId, clientHost,
+      rebalanceTimeoutMs, sessionTimeoutMs,
+      protocolType, List(("range", Array.empty[Byte]), ("roundrobin", Array.empty[Byte])))
+    member.assignment = ConsumerProtocol.serializeAssignment(
+      new PartitionAssignor.Assignment(util.Arrays.asList(tp3, tp4), null)).array
+    group.add(member)
+    group.currentStateTimestamp = None // Use the given timestamp instead
+
+    assert(Time.SYSTEM.milliseconds - now < retentionMs)
```

Comment:
Generally speaking, it's better to use mock timer instead of system time. There are examples in `GroupCoordinatorTest` which use mock timer.